torch/utils/checkpoint.py

```diff
@@ -0,0 +1,73 @@
+import torch
+from torch.autograd import Variable, Function
+
+
+def repackage_inputs(inputs, requires_grad=False):
+    if torch.is_tensor(inputs):
+        return Variable(inputs, requires_grad=requires_grad)
+    elif isinstance(inputs, tuple):
+        return tuple(repackage_inputs(v, requires_grad=requires_grad) for v in inputs)
+    else:
+        raise RuntimeError("Unknown input type")
+
+
+def unpack_variables(inputs):
+    if type(inputs) == Variable:
+        return inputs.data
+    elif torch.is_tensor(inputs):
+        return inputs
+    elif isinstance(inputs, tuple):
+        return tuple(unpack_variables(v) for v in inputs)
+    else:
+        raise RuntimeError("Unknown input type")
+
+
+class CheckpointFunction(Function):
+
+    # NOTE: *args is the flat inputs list, args is the tuple containing inputs
+    @staticmethod
+    def forward(ctx, run_function, *args):
+        ctx.run_function = run_function
+        ctx.save_for_backward(*args)
+        inputs = repackage_inputs(args)
+        with torch.no_grad():
+            outputs = run_function(*inputs)     # the *inputs* is always a tuple
+        return unpack_variables(outputs)
+
+    @staticmethod
+    def backward(ctx, *args):
+        saved_inputs = ctx.saved_tensors
+        inputs = repackage_inputs(saved_inputs, requires_grad=True)
```

Comment:
BTW repackaging everything with `requires_grad` if some of the inputs didn't require grad will increase the cost, but their gradients will be discarded anyway. Also, it still goes around in-place correctness checking, and we can't do that